libs/providers/go-feature-flag/src/lib/go-feature-flag-provider.ts

```diff
@@ -9,40 +9,132 @@ import {
+  private readonly dataFlushInterval: number;
+  
+  // disableDataCollection set to true if you don't want to collect the usage of flags retrieved in the cache.
+  private readonly disableDataCollection: boolean;
 
   constructor(options: GoFeatureFlagProviderOptions) {
     this.timeout = options.timeout || 0; // default is 0 = no timeout
     this.endpoint = options.endpoint;
+    this.cacheTTL = options.flagCacheTTL !== undefined && options.flagCacheTTL !== 0 ? options.flagCacheTTL : 1000 * 60;
+    this.dataFlushInterval = options.dataFlushInterval || 1000 * 60;
+    this.disableDataCollection = options.disableDataCollection || false;
 
     // Add API key to the headers
     if (options.apiKey) {
       axios.defaults.headers.common['Authorization'] = `Bearer ${options.apiKey}`;
     }
+
+    if (!options.disableCache) {
+      const cacheSize = options.flagCacheSize !== undefined && options.flagCacheSize !== 0 ? options.flagCacheSize : 10000;
+      this.cache = new LRUCache({maxSize: cacheSize, sizeCalculation: () => 1});
+    }
+  }
+
+  /**
+   * initialize is called everytime the provider is instanced inside GO Feature Flag.
+   * It will start the background process for data collection to be able to run every X ms.
+   */
+  async initialize() {
+    if (!this.disableDataCollection) {
+      this.bgScheduler = setInterval(async () => await this.callGoffDataCollection(), this.dataFlushInterval)
+      this.dataCollectorBuffer = []
+    }
+  }
+
+  /**
+   * onClose is called everytime OpenFeature.Close() function is called.
+   * It will terminate gracefully the provider and ensure that all the data are send to the relay-proxy.
+   */
+  async onClose() {
+    if (this.cache !== undefined && this.bgScheduler !== undefined) {
+      // we stop the background task to call the data collector endpoint
+      clearInterval(this.bgScheduler);
+      // We call the data collector with what is still in the buffer.
+      await this.callGoffDataCollection()
+    }
+  }
+
+
+  /**
+   * callGoffDataCollection is a function called periodically to send the usage of the flag to the
+   * central service in charge of collecting the data.
+   */
+  async callGoffDataCollection() {
+    if (this.dataCollectorBuffer?.length === 0) {
+      return
+    }
+
+    const dataToSend = structuredClone(this.dataCollectorBuffer);
+    this.dataCollectorBuffer = []
+
+    const request: DataCollectorRequest<boolean> = {events: dataToSend, meta: this.dataCollectorMetadata,}
+    const endpointURL = new URL(this.endpoint);
+    endpointURL.pathname = 'v1/data/collector';
+
+    try {
+      await axios.post<DataCollectorResponse>(endpointURL.toString(), request, {
+        headers: {
+          'Content-Type': 'application/json',
+          Accept: 'application/json',
+        },
+        timeout: this.timeout,
+      });
+    } catch (e) {
+      // TODO : add a log here
```

Comment:
Do you want to take care of this todo?

Reply:
Oh I forgot this one, I will look at it and log something in that case.